internal/lifecycle/readme.go

```diff
@@ -0,0 +1,149 @@
+	file, err := os.Open(filename)
+	if err != nil {
+		return nil, fmt.Errorf("[lifecycle.parseREADME] os.Open %s: %w", filename, err)
+	}
+	defer file.Close()
+
+	scanner := bufio.NewScanner(file)
+	var lifecycle Lifecycle
+	for scanner.Scan() {
+		line := scanner.Text()
+
+		if strings.Contains(line, codeTag) {
+			s := scanner.Scan()
+			startCodeBlockLine := scanner.Text()
+
+			c := strings.Contains(startCodeBlockLine, "```")
+			if !s || !c {
+				if err := scanner.Err(); err != nil && !s {
+					return nil, fmt.Errorf("[lifecycle.parseREADME] README bufio.Scanner: %w", err)
+				}
+
+				if !c {
+					return nil, fmt.Errorf("[lifecycle.parseREADME] parsing README: expecting start of code block immediately after code tag")
+				} else { // scanner.Scan falied, but no error found. EOF.
+					return nil, fmt.Errorf("[lifecycle.parseREADME] parsing README: unexpected EOF; file ended immediately after code tag")
+				}
+			}
+
+			var blockClosed bool
+			for scanner.Scan() {
+				line = scanner.Text()
+				if strings.Contains(line, "```") {
+					blockClosed = true
+					break
+				}
+
+				line = strings.TrimSpace(line)
+
+				// If there is a backslash at the end of the line, this is a multiline command. Keep scanning to get
+				// entire command.
+				for line[len(line)-1] == '\\' {
+					line = line[:len(line)-1]
+
+					scanner.Scan()
+					line = line + strings.TrimSpace(scanner.Text())
+				}
+
+				line = os.ExpandEnv(line)
+				line = replaceGCRURL(line, gcrURL)
+				line = replaceServiceName(line, serviceName)
+				sp := strings.Split(line, " ")
+
+				var cmd *exec.Cmd
+				if strings.Contains(line, "gcloud") {
+					a := append(util.GcloudCommonFlags, sp[1:]...)
+					cmd = exec.Command("gcloud", a...)
+				} else {
+					cmd = exec.Command(sp[0], sp[1:]...)
+				}
+
+				lifecycle = append(lifecycle, cmd)
+			}
+
+			if err := scanner.Err(); err != nil {
+				return nil, fmt.Errorf("[lifecycle.parseREADME] README bufio.Scanner: %w", err)
+			}
+
+			if !blockClosed {
+				return nil, fmt.Errorf("[lifecycle.parseREADME] parsing README: code block not closed before end of file")
+			}
+		}
+	}
+
+	if err := scanner.Err(); err != nil {
+		return nil, fmt.Errorf("[lifecycle.parseREADME] README bufio.Scanner: %w", err)
+	}
+
+	if len(lifecycle) == 0 {
+		return nil, fmt.Errorf("[lifecycle.parseREADME] parsing README: no commands found")
+	}
+
+	return lifecycle, nil
+}
+
+// replaceServiceName takes a terminal command string as input and replaces the Cloud Run service name, if any.
+// It detects whether the command is a gcloud run command and replaces the last argument that isn't a flag
+// with the input service name.
+func replaceServiceName(command, serviceName string) string {
+	r1 := regexp.MustCompile(`\bgcloud\b`)
+	r2 := regexp.MustCompile(`\brun\b`)
+
+	containsGcloud := r1.MatchString(command)
+	containsRun := r2.MatchString(command)
+
+	if !(containsGcloud && containsRun) {
```

Comment:
nit: You can avoid adding these variables to scope as well
```suggestion
	if !(containsGcloud.MatchString(command) && containsRun.MatchString(command)) {
```

Reply:
Resolved, thank you.